### ISSUE TITLE
Duplicate absMid for some modules

### ISSUE THREAD

**Issue by cmdcolin** [OpenNTF/dojo-webpack-plugin#211]
I have a module that uses babel-runtime (6) and @babel/runtime (7) and it was running into some issues with duplicate absMid

```
Duplicate absMid ("core-js/library/fn/number/max-safe-integer") for modules "/home/cdiesh/src/jbrowse/node_modules/@babel/runtime-corejs2/node_modules/core-js/library/fn/number/max-safe-integer.js" and "/home/cdiesh/src/jbrowse/node_modules/babel-runtime/node_modules/core-js/library/fn/number/max-safe-integer.js"
```

Is it possible to work around this bug? I also noticed that I was getting duplicate absMid when I did "yarn link" setups similar to this, but this is the first time I noticed the bug that is unresolvable due to mixing babel 6 and 7.

Currently using 2.7.7, can try and make a reproducible example if it helps

**Comment by chuckdumont** [comment]
Hmm.  You shouldn't be getting that error for CommonJS modules.  I'll look into it.

**Comment by cmdcolin** [comment]
Also I think this might be related to using webpack 4, I forgot to mention that...at least the yarn link seemed to work better before then and then had trouble after that

Sorry about the lack of small reproducible example, let me know if that would be of a big help to fix it

**Comment by cmdcolin** [comment]
If you want a non-small reproducible example you can try using yarn and then yarn watch with this https://github.com/GMOD/jbrowse/tree/bump_tabix

**Comment by chuckdumont** [comment]
@cmdcolin - please try with latest version of dojo-webpack-plugin.

**Comment by cmdcolin** [comment]
I tried with 2.8.3 and it gives the same error

**Comment by GordonSmith** [comment]
FWIW I am also seeing this issue with v2.8.3 - but not with v2.7.6

**Comment by chuckdumont** [comment]
@cmdcolin, @GordonSmith - please verify that v2.8.4-0 resolves the problem.

**Comment by GordonSmith** [comment]
It doesn't appear to:
* I `npm install` v2.8.3 (as v2.8.4 is not in npmjs yet)
* Checked out v2.8.4-0 to local folder
* Ran npm install
* Ran npm pack
* Copied the contents of dojo-webpack-plugin-2.8.4-0.tgz over my existing 2.8.3 in node_modules

Here is my output:
```

ERROR in chunk index [entry]
Duplicate absMid ("dgrid/extensions/nls/pagination") for modules "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js" and "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dojo-webpack-plugin\\loaders\\dojo\\i18nRootModifier\\index.js?bundledLocales=!C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js"

ERROR in chunk index.min [entry]
Duplicate absMid ("dgrid/extensions/nls/pagination") for modules "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js" and "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dojo-webpack-plugin\\loaders\\dojo\\i18nRootModifier\\index.js?bundledLocales=!C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js"

ERROR in chunk index [entry]
index.js
Duplicate absMid ("dgrid/extensions/nls/pagination") for modules "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js" and "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dojo-webpack-plugin\\loaders\\dojo\\i18nRootModifier\\index.js?bundledLocales=!C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js"

ERROR in chunk index.min [entry]
index.min.js
Duplicate absMid ("dgrid/extensions/nls/pagination") for modules "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js" and "C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dojo-webpack-plugin\\loaders\\dojo\\i18nRootModifier\\index.js?bundledLocales=!C:\\Users\\gordon\\git\\hpcc-js\\node_modules\\dgrid\\extensions\\nls\\pagination.js"
```

**Comment by cmdcolin** [comment]
@chuckdumont I think this actually worked for me, tested on ubuntu, node 11.11, with dojo webpack 2.8.4-0 (cc @GordonSmith  I just used yarn add -D dojo-webpack-plugin@next)

**Comment by GordonSmith** [comment]
FYI - I set my NPM version to 2.8.4-0 (which worked), but I am seeing the same errors I posted above.
Steps to reproduce (large mono repo I am afraid):
```
git clone https://github.com/hpcc-systems/Visualization.git
cd Visusalization
git checkout DOJO_WEBPACK
npm install
cd packages
cd dgrid-shim
npm run build
```

**Comment by chuckdumont** [comment]
@GordonSmith - here's what's going on with your build.  In order to support the [locales](https://github.com/OpenNTF/dojo-webpack-plugin#locales) option, the dojo/i18n webpack loader (part of dojo-webpack-plugin) edits root nls bundles to remove excluded locales.  For example, if you specify `locals:["en", "es"]` in the plugin options and your nls bundle looks like the following:

```javascript
define({ root:
	({
		hello: "Hello"
	}),
	"es": true,
	"fr": true
})
```

then the root bundle will be modified to remove excluded locales as follows:
```javascript
define({ root:
	({
		hello: "Hello"
	}),
	"es": true
})
```

This is done using the dojo/i18nRootModifier webpack plugin (also part of dojo-webpack-plugin) and normally does not present a problem.  In your case, however, your code appears to be loading dgrid/extensions/nls/pagination both with dojo/i18n and as a normal module.  This results in the same absMid, dgrid/extensions/nls/pagination, referring to two different modules with different content.

There are a few ways you can deal with this:

- Always use the dojo/i18n plugin to load dojo nls bundles.
- Include all languages in your webpack build (`locales: '*'`)
- Use the webpack NormalModuleReplacementPlugin to ensure that if nls modules are loaded without the use of the dojo/i18n plugin, then they will still be processed by the dojo/i18nRootModifier plugin.  In your case, that would look something like the following:
```javascript
      new webpack.NormalModuleReplacementPlugin(
        /^dgrid/extensions/\/nls\/[^/]*$/, function(data) {
          data.request = "dojo/i18nRootModifier?bundledLocales=!" + data.request;
        }
      )
``` 
Note that the bundledLocales query arg delimits multiple locales using the | character and when invoked by the dojo/i18n plugin, does not include locales which are not enabled in the bundle being modified.

**Comment by chuckdumont** [comment]
@GordonSmith - I've changed the way that the dojo/i18nRootModifier plugin is invoked.  It is no longer invoked explicitly by the dojo/i18n plugin.  Instead, it is now invoked by webpack based on filename pattern matching.  This should resolve your issue without the need for any changes to your code or your config.  Please try out version 2.8.4-1.  Thanks.

**Comment by chuckdumont** [comment]
These issues are resolved in release 2.8.5.
